File: src-tauri/src/utils/step_events.rs
use serde::Serialize;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use tauri::Emitter;

use crate::utils::logger_utils::global_app_handle;

const STEP_EVENT: &str = "launch-steps";
const INTEGRITY_EVENT: &str = "integrity-steps";

#[derive(Clone, Copy)]
pub enum StepChannel {
    Launch,
    Integrity,
}

impl StepChannel {
    fn event_name(&self) -> &'static str {
        match self {
            StepChannel::Launch => STEP_EVENT,
            StepChannel::Integrity => INTEGRITY_EVENT,
        }
    }
}

#[derive(Serialize, Clone)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum StepEvent {
    Started { id: String, label: String },
    Progress { id: String, current: u64, total: u64 },
    Detail { id: String, text: String },
    Finished { id: String, skipped: bool },
    Failed { id: String, message: String },
}

fn emit_step_event(channel: StepChannel, event: &StepEvent) {
    if let Some(app) = global_app_handle() {
        let _ = app.emit(channel.event_name(), event);
    }
}

#[derive(Clone)]
pub struct StepHandle {
    id: &'static str,
    state: Arc<StepState>,
    channel: StepChannel,
}

struct StepState {
    current: AtomicU64,
    total: AtomicU64,
}

impl StepHandle {
    pub fn start(id: &'static str, label: impl Into<String>) -> Self {
        Self::start_channel(StepChannel::Launch, id, label)
    }

    pub fn start_channel(channel: StepChannel, id: &'static str, label: impl Into<String>) -> Self {
        emit_step_event(
            channel,
            &StepEvent::Started {
                id: id.to_string(),
                label: label.into(),
            },
        );
        Self {
            id,
            state: Arc::new(StepState {
                current: AtomicU64::new(0),
                total: AtomicU64::new(0),
            }),
            channel,
        }
    }

    pub fn detail(&self, text: &str) {
        emit_step_event(
            self.channel,
            &StepEvent::Detail {
                id: self.id.to_string(),
                text: text.to_string(),
            },
        );
    }

    pub fn set_total(&self, total: u64) {
        self.state.current.store(0, Ordering::Relaxed);
        self.state.total.store(total, Ordering::Relaxed);
        emit_step_event(
            self.channel,
            &StepEvent::Progress {
                id: self.id.to_string(),
                current: 0,
                total,
            },
        );
    }

    pub fn inc(&self) {
        let current = self.state.current.fetch_add(1, Ordering::Relaxed) + 1;
        let total = self.state.total.load(Ordering::Relaxed);
        emit_step_event(
            self.channel,
            &StepEvent::Progress {
                id: self.id.to_string(),
                current,
                total,
            },
        );
    }

    pub fn finish(self, skipped: bool) {
        emit_step_event(
            self.channel,
            &StepEvent::Finished {
                id: self.id.to_string(),
                skipped,
            },
        );
    }

    pub fn fail(self, message: impl Into<String>) {
        emit_step_event(
            self.channel,
            &StepEvent::Failed {
                id: self.id.to_string(),
                message: message.into(),
            },
        );
    }
}

#[macro_export]
macro_rules! step_try {
    ($step:expr, $expr:expr) => {
        match $expr {
            Ok(value) => value,
            Err(error) => {
                $step.fail(error.to_string());
                return Err(error.into());
            }
        }
    };
}

File: src/06-shared/utils/stepEvents.ts
import type { StepEvent, StepProgressItem } from '@/05-entities'

export function createStepItem(key: string, label: string, shownAt: number): StepProgressItem {
  return {
    key,
    label,
    status: 'active',
    skipped: false,
    current: 0,
    total: 0,
    detail: '',
    error: '',
    shownAt,
  }
}

export function applyStepEvent(steps: StepProgressItem[], event: StepEvent): void {
  switch (event.type) {
    case 'started': {
      const active = steps.find((step) => step.status === 'active')
      if (active) active.status = 'done'
      const index = steps.findIndex((item) => item.key === event.id)
      if (index === -1) {
        steps.push(createStepItem(event.id, event.label, Date.now()))
        break
      }
      const step = steps[index]
      if (step) {
        step.status = 'active'
        step.label = event.label
        step.shownAt = Date.now()
      }
      for (let i = 0; i < index; i += 1) {
        const before = steps[i]
        if (before !== undefined && before.status === 'pending') {
          before.status = 'done'
          before.skipped = true
        }
      }
      break
    }
    case 'progress': {
      const step = steps.find((item) => item.key === event.id)
      if (step) {
        step.current = event.current
        step.total = event.total
      }
      break
    }
    case 'detail': {
      const step = steps.find((item) => item.key === event.id)
      if (step) step.detail = event.text
      break
    }
    case 'finished': {
      const step = steps.find((item) => item.key === event.id)
      if (step) {
        step.status = 'done'
        step.skipped = event.skipped
        step.detail = ''
      }
      break
    }
    case 'failed': {
      const step = steps.find((item) => item.key === event.id)
      if (step) {
        step.status = 'error'
        step.error = event.message
        step.detail = ''
      }
      break
    }
  }
}

export function computeStepProgress(steps: StepProgressItem[]): number {
  if (steps.length === 0) return 0
  let done = 0
  let fraction = 0
  for (const step of steps) {
    if (step.status === 'done' || step.status === 'error') {
      done++
    } else if (step.status === 'active' && step.total > 0) {
      fraction = Math.min(step.current / step.total, 1)
    }
  }
  return ((done + fraction) / steps.length) * 100
}
